src/smallfry.c

```diff
@@ -521,7 +523,7 @@ float metric_nhw (uint8_t *inbuf, uint8_t *outbuf, int width, int height)
             resd = abs(res - resr);
             if (resd > 0)
             {
-                delta = abs((float)new[scan] - (float)old[scan]);
+                delta = abs(new[scan] - old[scan]);
```

Comment:
Implicit cast.
```C
#include <stdlib.h>
#include <stdio.h>
#include <stdint.h>

int main() {
  float f = 0.0f;
  int n = 0;
  uint8_t a = 0;
  uint8_t b = 250;
  f = abs(a - b);
  n = a - b;
  printf("f = %f\n", f);
  printf("n = %d\n", n);
}
```

Output.
```
$ gcc test.c -o test -lm

$ ./test
f = 250.000000
n = -250
```

Reply:
Optimizations for this are sometimes given by UB. Maybe?:
```c
delta = (new[scan] < old[scan]) ? (old[scan] - new[scan]) : (new[scan] - old[scan]);
```

Reply:
Would be better to explicitly cast to int than add branching instructions.  However, there is no undefined behavior.  It is specified in [C99 Language Standard 6.3.1.1](https://rgambord.github.io/c99-doc/sections/6/3/1/1/index.html):

> If an `int` can represent all values of the original type, the value is converted to an `int`; otherwise, it is converted to an `unsigned int`.  These are called the integer promotions.

Section [6.3.1.8](https://rgambord.github.io/c99-doc/sections/6/3/1/8/index.html) reiterates:

> ... the integer promotions are performed on both operands...

Unless this is running on an 8-bit processor, I doubt any optimization would affect the result.

Reply:
ok. Give tests for the identity of function outputs on a real image.